### PR TITLE
PLANNER-350: Workbench: scoreHolder support in Guided Rule Editor

### PR DESCRIPTION
**Pull request by mcimbora**
Part of:
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/425
https://github.com/kiegroup/drools/pull/1178
https://github.com/kiegroup/drools-wb/pull/433
https://github.com/kiegroup/optaplanner-wb/pull/143
https://github.com/kiegroup/kie-wb-distributions/pull/486

### CODE REVIEW

**Comment by manstis** [comment]
@mcimbora Hi, could you please format code with the latest (InteliJ) formatter in ```droolsjbpm-build-bootstrap```? I see you've added back a load of white space previously removed when the latest formatter is used.

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java [hunk 1]
```diff
@@ -172,11 +174,20 @@ public static RuleModelPersistence getInstance() {
      * org.drools.ide.common.server.util.RuleModelPersistence#marshal(org.drools.guvnor
      * .client.modeldriven.brl.RuleModel)
      */
-    public String marshal(final RuleModel model) {
-        return marshalRule(model);
+    public String marshal( final RuleModel model ) {
+        return marshalRule( model,
+                            null );
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java]: ```Collections.emptySet()``` (or similar) would probably be nicer?

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java [hunk 2]
```diff
@@ -429,8 +443,19 @@ protected void marshalRHS(final StringBuilder buf,
             }
 
             //Now generate DRL
-            for (IAction action : model.rhs) {
-                actionVisitor.visit(action);
+            for ( IAction action : model.rhs ) {
+                if ( extensions != null ) {
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java]: Use of ```Optional``` or guarantee of a non-empty collection would be nicer than ```null``` checks.

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java [hunk 3]
```diff
@@ -429,8 +443,19 @@ protected void marshalRHS(final StringBuilder buf,
...
+                    List<RuleModelIActionPersistenceExtension> matchingExtensions = extensions.stream().filter( e -> e.accept( action ) ).collect( Collectors.toList() );
+                    if ( matchingExtensions.isEmpty() ) {
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java]: Sure the ```stream``` API could _possibly_ be better used (```stream().filter(..).findFirst().orElseGet(..)```)

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java [hunk 4]
```diff
@@ -1830,39 +1860,59 @@ private void getClasses(ActionFieldValue[] fieldValues) {
...
      */
-    public RuleModel unmarshal(final String str,
-                               final List<String> globals,
-                               final PackageDataModelOracle dmo) {
-        if (str == null || str.isEmpty()) {
+    public RuleModel unmarshal( final String str,
+                                final List<String> globals,
+                                final PackageDataModelOracle dmo ) {
+
+        return unmarshal( str, globals, dmo, null );
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java]: ```Collections.emptySet()``` or use of ```Optional``` would be preferential.

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java [hunk 5]
```diff
@@ -1830,39 +1860,59 @@ private void getClasses(ActionFieldValue[] fieldValues) {
...
-    public RuleModel unmarshalUsingDSL(final String str,
-                                       final List<String> globals,
-                                       final PackageDataModelOracle dmo,
-                                       final String... dsls) {
-        if (str == null || str.isEmpty()) {
+    public RuleModel unmarshalUsingDSL( final String str,
+                                        final List<String> globals,
+                                        final PackageDataModelOracle dmo,
+                                        final String... dsls ) {
+        return unmarshalUsingDSL( str, globals, dmo, null, dsls );
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java]: Use of ```Collections.emptySet()``` or ```Optional``` would be preferential.

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java [hunk 6]
```diff
@@ -2741,18 +2794,44 @@ private void parseRhs(final RuleModel m,
...
             lineCounter++;
-            if (expandedDRLInfo.hasDsl) {
-                String dslLine = expandedDRLInfo.dslStatementsInRhs.get(lineCounter);
-                while (dslLine != null) {
-                    m.addRhsItem(toDSLSentence(expandedDRLInfo.rhsDslPatterns,
-                                               dslLine));
-                    dslLine = expandedDRLInfo.dslStatementsInRhs.get(++lineCounter);
+            line = line.trim();
+
+            if ( extensions != null ) {
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java]: Use of ```Optional``` (but I think you get the message now.....) ;-)

**Review by manstis — COMMENTED** [review]
General comment about use of ```Optional``` (or anything better than ```null```).. otherwise approach seems reasonable (w/o reviewing other PRs yet....)

**Comment by mcimbora** [comment]
Jenkins retest this.

File: drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelIActionPersistenceExtension.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package org.drools.workbench.models.commons.backend.rule;
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelIActionPersistenceExtension.java:17]: License please.

**Review by manstis — COMMENTED** [review]
@mcimbora Just a missing license then this is good to go!

**Comment by mcimbora** [comment]
@manstis Thanks Mike, PR updated.

**Review by manstis — APPROVED** [review]

**Pull request merged**